src/backend/catalog/pg_directory_table.c

```diff
@@ -0,0 +1,181 @@
+typedef struct DirTableColumnDesc
+{
+	const char *colName;
+	const char *typName;
+} DirTableColumnDesc;
+
+static const DirTableColumnDesc dirTableColumns[] = {
+	{"relative_path", "text"},
+	{"size", "int8"},
+	{"last_modified", "timestamptz"},
+	{"md5", "text"},
+	{"tag", "text"}
+};
+
+static void
+GetTablespaceFileHandler(Oid spcId)
+{
+	HeapTuple tuple;
+	Datum datum;
+	bool isNull;
+	Oid	fileHandlerOid;
+	Form_pg_tablespace tblspcForm;
+
+	tuple = SearchSysCache1(TABLESPACEOID, ObjectIdGetDatum(spcId));
+	if (!HeapTupleIsValid(tuple))
+		elog(ERROR, "cache lookup failed for table space %u", spcId);
+
+
+	tblspcForm = (Form_pg_tablespace) GETSTRUCT(tuple);
+	datum = SysCacheGetAttr(TABLESPACEOID,
+							tuple,
+							Anum_pg_tablespace_spcfilehandler,
+							&isNull);
+	if (!isNull)
+	{
+		fileHandlerOid = DatumGetObjectId(datum);
+		datum = OidFunctionCall0(fileHandlerOid);
+		currentFileAm = (FileAm *) DatumGetPointer(datum);
+		if (currentFileAm == NULL)
+			elog(ERROR, "tablespace file handler %u did not return a FileAm struct",
+				 		fileHandlerOid);
+	}
+	else
+	{
+		currentFileAm = &localFileAm;
+	}
+
+	ReleaseSysCache(tuple);
+
+	Assert(currentFileAm != NULL);
+	Assert(currentFileAm->close != NULL);
+	Assert(currentFileAm->read != NULL);
+	Assert(currentFileAm->write != NULL);
+	Assert(currentFileAm->size != NULL);
+	Assert(currentFileAm->name != NULL);
+	Assert(currentFileAm->getLastError != NULL);
+
+	return;
+}
+
+/*
+ * GetDirectoryTable - look up the directory table definition by relId.
+ */
+DirectoryTable *
+GetDirectoryTable(Oid relId)
+{
+	Form_pg_directory_table dirtableForm;
+	HeapTuple tuple;
+	Datum datum;
+	bool isNull;
+	DirectoryTable *dirTable;
+
+	tuple = SearchSysCache1(DIRECTORYTABLEREL, ObjectIdGetDatum(relId));
+	if (!HeapTupleIsValid(tuple))
+		elog(ERROR, "cache lookup failed for directory table %u", relId);
+
+	dirtableForm = (Form_pg_directory_table) GETSTRUCT(tuple);
+
+	dirTable = (DirectoryTable *) palloc(sizeof(DirectoryTable));
+	dirTable->relId = relId;
+	dirTable->spcId = dirtableForm->dttablespace;
+	GetTablespaceFileHandler(dirtableForm->dttablespace);
+
+	datum = SysCacheGetAttr(DIRECTORYTABLEREL,
+							tuple,
+							Anum_pg_directory_table_dtlocation,
+							&isNull);
+	Assert(!isNull);
+
+	dirTable->location = TextDatumGetCString(datum);
+	ReleaseSysCache(tuple);
+
+	return dirTable;
+}
+
+bool
+RelationIsDirectoryTable(Oid relId)
+{
+	HeapTuple tuple;
+
+	tuple = SearchSysCache1(DIRECTORYTABLEREL, ObjectIdGetDatum(relId));
+	if (!HeapTupleIsValid(tuple))
+		return false;
+
+	ReleaseSysCache(tuple);
+	return true;
+}
+
+List *
+GetDirectoryTableBuiltinColumns(void)
+{
+	int i;
+	List *result = NIL;
+
+	for (i = 0; i < lengthof(dirTableColumns); i++)
+	{
+		ColumnDef *columnDef = makeNode(ColumnDef);
+
+		columnDef->colname = pstrdup(dirTableColumns[i].colName);
+		columnDef->typeName = SystemTypeName(pstrdup(dirTableColumns[i].typName));
+		columnDef->is_local = true;
+
+		if (i == 0)
```

Comment:
only i = 0 need constraint ?

Reply:
Yes, the directory table schema table need to build primary key on the first column